### ISSUE TITLE
[BUG] Auto multi-line regular expression matches too broadly

### ISSUE THREAD

**Issue by ollien** [DataDog/datadog-agent#16349]
**Agent Environment**
N/A - this was found by code inspection in `main`

**Describe what happened:**
There is a list of regular expressions used for multi-line auto detection. I was validating my log formats against this list and noticed this

https://github.com/DataDog/datadog-agent/blob/70396ecb61bb6aa179c6b34aa842597d1aa3b865/pkg/logs/internal/decoder/auto_multiline_handler.go#L224C4-L225
```
	// 2021-01-31 - with stricter matching around the months/days
	regexp.MustCompile(`^\d{4}-(0?[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])`),
```

This pattern does not match the example given

https://regex101.com/r/lQSO1e/1

It seems that the problem is that the parser sees that there is no `0` at the start of the day, and thne matches the next digit, at which point the match is complete.

**Describe what you expected:**
I would expect the given pattern to match `2021-01-31`. I think the correct fix is to move the `0?[1-9]` alternative to the end of the list like so `^\d{4}-(1[012]|0?[1-9])-([12][0-9]|3[01]|0?[1-9])`. This forces the regular expression parser to consider the fixed-width groups first.

https://regex101.com/r/FXlCxl/1

**Steps to reproduce the issue:**

See above regex101s

**Additional environment details (Operating System, Cloud provider, etc):**

**Comment by gh123man** [comment]
Thanks for raising this. Your fix is correct, so I've made a PR here: https://github.com/DataDog/datadog-agent/pull/16432

**Comment by carlosroman** [comment]
Closing as fixed by #16432
